add Changeset APIs + CLI exporting

vito opened the pull request:
Introduces `Changeset`, a new API representing a diff between two directories - similar to `Directory.diff`, except it can represent _removals_ as well.

[Schema diff](https://github.com/dagger/dagger/pull/10946/files#diff-6d45938d3a99ad730e8d639cd2218f57c90828f1deeaad17c073cad22bfd0b50)

## Better export mechanics

A goal of this PR is to address #8235. The plan is:

* [x] When the CLI calls a function that returns `Changes`, it prompts to automatically apply them to your working directory, after showing you a diff.
* Following #10907 - when `Changes` are returned, they are automatically applied to the `Env` hostfs.

Pros:

* It's very clear now when an API returns a full `Directory` or a delta to be applied.
* We don't need to wipe the entire destination anymore, now that we know exactly what paths were removed.
* A single API returning `Changeset` can be used both as a CLI entrypoint and as part of the internal build pipeline via `Directory.withChanges`.

## Future possibilities

An API like this would be neat, and dovetail nicely into the new `Directory.withPatch` API:

```graphql
extend type Changeset {
  "Returns a Git-compatible patch file."
  asPatch: File!
}
```

Building on this, we could potentially have APIs for `git` committing changes etc., possibly using the patch format as a generic medium.

shykes commented:
Nit: I would call it `Change`, singular. There's a pretty common convention that a type name that is plural designates an array or some sort of collection. Better to avoid that confusion IMO.

shykes commented:
> ```graphql
> extend type Directory {
>   """
>   Return a virtual comparison between this directory and an older snapshot that can be applied to another filesystem.
>   """
>   changes(
>     """The older directory snapshot to compare against"""
>     older: DirectoryID!
>   ): Changes!

This deprecates `diff()`, right? I would either A) straight up break `diff` (assuming we can use compat mode + make migration easy), or B) call it something obviously similar, like `difference` or `compare`. I think `changes()` is too vague IMO - also not really symmetrical to `withChanges()`, so the `X() / withX()` pair doesn't really work IMO.

>   """Return a directory with changes from another directory applied to it."""
>   withChanges(
>     """Changes to apply to the directory"""
>     changes: ChangesID!
>   ): Directory!
> }

Per my previous comment on Change singular, I'd do:

```graphql
  withChange(change: ChangeID)
```

vito commented:
@shykes 

> Nit: I would call it `Change`, singular. There's a pretty common convention that a type name that is plural designates an array or some sort of collection. Better to avoid that confusion IMO.

Hmm, yeah that's true at least in some languages (like Go), and we kinda have to avoid friction across all of them.

`Change` feels weird to me though. To me it _does_ feel like it should be plural, considering it represents a set of changed files, not just one change. Depends on where you draw the lines in your head - is a Git commit one change, or many changes?

* Ironically we also have `File.contents`, which I always thought should be `File.content` - maybe that's why I went for `Changes` :smile:  

Maybe we can leapfrog this whole issue by calling it `Diff` instead, or maybe even `DirectoryDiff` to be more conservative.

Or `Patch`?

---

> This deprecates `diff()`, right? I would either A) straight up break `diff` (assuming we can use compat mode + make migration easy), or B) call it something obviously similar, like `difference` or `compare`. I think `changes()` is too vague IMO

The `diff` migration is a little fraught, since I also changed the order of the arguments, because I got it wrong every single time I used it. I always wanted to keep the chain going, like `foo.bar.baz.diff(foo)`, but actually you have to pre-medidate the diff, like `foo.diff(foo.bar.baz)`.

I went with "changes" because it sounded a little more directional -- you're asking the directory that _has_ the changes, what they are relative to some other directory. Not married to it though. Words are hard.

`changesFrom`? `diffFrom`?

>  - also not really symmetrical to `withChanges()`, so the `X() / withX()` pair doesn't really work IMO.

FWIW the symmetry seems the same to me:

```
c = dir.changes(olderDir)
dir.withChanges(c) will always == dir

d = dir.directory("foo")
dir.withDirectory("foo", d) will always == dir
```

jedevc commented:
Continuing bikeshedding...

I'm not a huge `Changes` fan either, `ChangeSet` feels a bit *strange* to me though coming from a git-world, where it's usually a *series* of changes (i.e. multiple commits) - this is just one `Change` (which if we're going for `Change*` is probably my preference).

`Diff` would be nice, and probably my favorite! That also feels nice going into #7202, you could just do `GitRef.newCommit(diff: Diff, ...)` or whatever. Also, `Diff.asPatch` feels cool!

Frustrating that `Directory.diff` then returns a `Directory`. I'd happily see that functionality updated, so that it returns a `Diff` - we'd need version compat for it, and probably would still need to keep an internal `__diff` or something so that we can keep the functionality for the `Select` call.

shykes commented:
> `ChangeSet` feels a bit _strange_ to me though coming from a git-world, where it's usually a _series_ of changes (i.e. multiple commits) - this is just one `Change` (which if we're going for `Change*` is probably my preference).

In git terminology "changeset" usually refers to the set of changes in a single commit.

> `Diff` would be nice, and probably my favorite! That also feels nice going into #7202, you could just do `GitRef.newCommit(diff: Diff, ...)` or whatever. Also, `Diff.asPatch` feels cool!

Yes `Diff` also works, but it's weird that `Directory.diff()` doesn't return a `Diff`...

> Frustrating that `Directory.diff` then returns a `Directory`.

Haha exactly :)

> I'd happily see that functionality updated, so that it returns a `Diff` - we'd need version compat for it, and probably would still need to keep an internal `__diff` or something so that we can keep the functionality for the `Select` call.

Yeah breaking `Directory.diff()` seems like a can of worms, per @vito.

vito commented:
I'm good with `ChangeSet`, with a slight preference for `Changeset` (per [Wikipedia](https://en.wikipedia.org/wiki/Changeset)) to make it clear that it's a discrete type and just a Set of Changes (which doesn't have its own type).

Tangent: to keep pulling on this Git/VCS analogy: what if you could have an agent look at PR reviews, start resolving all feedback in parallel, and then attempt to merge all of those `Changeset`s together? That would mean we actually have to do file content-wise merge, not just "last file layer wins" which is what you'd get without any special effort.

And then how do we handle conflicts? Maybe you could give that to another agent to resolve, or bubble it up to the user? (The latter is probably safer, last time I tried to tell an agent to resolve merge conflicts it just blindly picked one side, but maybe one day... or maybe with enough context.) In that case we'd probably want a `conflictedPaths` getter.

That can all be future stuff though if we want it. Just thinking out loud.

vito commented on core/directory.go:
```diff
@@ -930,6 +924,189 @@ func (dir *Directory) Diff(ctx context.Context, other *Directory) (*Directory, e
+
+type Changes struct {
+	Before dagql.ObjectResult[*Directory] `field:"true" doc:"The older/lower snapshot to compare against."`
+	After  dagql.ObjectResult[*Directory] `field:"true" doc:"The newer/upper snapshot."`
```
Yeah sounds good to me - especially if we plan to deprecate `diff` we'll need a way to actually, you know, get the diff. :smile: 

I'm not sure what to call it though. It's not really `diff` because it's a `Directory` and the problem this new type is solving is that a `Directory` can't represent removals.

Maybe `writes`? (Or is a removal still considered a "write"...)

Request For Dictionary @shykes

shykes replied:
I would call it `layer` or `asLayer`.

shykes reviewed (approved):
Small API nits, otherwise looks great!

I found a type name I like better than `Changeset`: `Delta` :)

vito commented:
Addressed the outstanding TODO for having the CLI apply changes when a function returns a `Changeset`:

[![asciicast](https://asciinema.org/a/737614.svg)](https://asciinema.org/a/737614)

Which involved adding some more APIs:

* `Changeset.layer`: as requested
* `Changeset.asPatch: File!`: returns the changes in a Git-compatible patch format, used for displaying the change summary
  * shares test coverage between `Directory.withChanges` and `Directory.withPatchFile(Changeset.asPatch)`
* `Changeset.sync`: in dogfooding I noticed a few places that call `.Sync` to pre-compute things in goroutines, so I just added it
  * ideally this sort of thing would become unnecessary by evaluating chained API call args in parallel in `dagql`
* `Directory.withPatchFile`: a `File`-accepting form of `withPatch`
* `Directory.withPatch` and `withPatchFile` now no-op with an empty patch, rather than failing

Along the way I switched to `charmbracelet/huh` for our TUI prompts which should be much nicer going forward as it's about a billion times more flexible and powerful than our current prompt implementation.

@shykes @jedevc Sorry for expanding scope post-approval, but that last TODO felt important; if these changes are too egregious I can split them out to another PR, but my understanding is nothing is blocked on `Changeset` so might as well get it as good as possible before merging. If we're not good-to-go prior to vacation I'm happy to hand the PR off for final touches etc.

jedevc commented on core/schema/wrapper.go:
```diff
@@ -159,6 +159,9 @@ func getSelfDigest(a any) (digest.Digest, error) {
 	case *core.GitRef:
 		// FIXME can core.DigestOf(x) be used instead? When set, the TestGit/TestAuthClient test failed intermittently
 		return "", nil // fallback to using dagop ID
+	case *core.Changeset:
+		// FIXME this seems weird?
+		return "", nil
```
Indeed :sweat_smile: 

Some discussion in private discord here: https://discord.com/channels/707636530424053791/1347633524093616190/1416099558315786451

I wonder how much of this might be removable at some point :thinking:

jedevc commented on cmd/dagger/functions.go:
```diff
@@ -694,6 +702,207 @@ func handleResponse(returnType *modTypeDef, response any, o, e io.Writer) error
+		return err
+	}
+
+	removed, err := changeset.RemovedPaths(ctx)
+	if err != nil {
+		return err
+	}
```
Question? If I added `Host.removeFiles` or similar, would this be able to be moved to the engine?

Dunno if it's a huuuge blocker, but I can't see if it's possible to export a changeset from `dagger shell`.

jedevc commented on cmd/dagger/functions.go:
```diff
@@ -694,6 +702,207 @@ func handleResponse(returnType *modTypeDef, response any, o, e io.Writer) error
+	return nil
+}
+
+func summarizePatch(out *termenv.Output, patch string, maxWidth int) error {
```
Move to `idtui`?

jedevc commented on core/integration/directory_test.go:
```diff
@@ -755,6 +755,863 @@ func (DirectorySuite) TestDiff(ctx context.Context, t *testctx.T) {
 	*/
 }
 
+func (DirectorySuite) TestChanges(ctx context.Context, t *testctx.T) {
```
nit: maybe a separate file for these tests?

jedevc commented on core/integration/directory_test.go:
Can't see tests for the output flag.

vito replied:
Added!

jedevc commented on cmd/dagger/functions.go:
```diff
@@ -694,6 +702,207 @@ func handleResponse(returnType *modTypeDef, response any, o, e io.Writer) error
+		exportDest = "."
+
+		var confirm bool
+		form := idtui.NewForm(
```
Should we still prompt even if there are no changes?

<img width="708" height="281" alt="image" src="https://github.com/user-attachments/assets/014d390f-ef2e-4fe8-9254-22a39814ac45" />

This feels a bit weird, since it won't do anything if I hit `y`.

jedevc commented:
Neat! I like it, code looks good!

`huh` is a very neat little library :grin:

jedevc reviewed (approved):
Approving, LGTM :tada: :tada:

Ping @shykes on the API, but IMO, they're good!

shykes commented:
@jedevc yup looks good, we already did some bikeshedding on the name, time to ship :)

(thanks for checking)
